client/collator/src/lib.rs

```diff
@@ -241,6 +241,40 @@ where
 				}
 			};
 
+			let code_hash = sp_io::storage::get(well_known_keys::CODE_HASH);
```

Comment:
IIRC this code is executed after the block is executed, thus the code is the code after runtime upgrade. Instead we need to fetch the code before runtime upgrade. So I need to find a way to get the code hash from the code used to build the code.

If not then I mistake somewhere else my manual test show that the produced candidate is considered invalid by polkadot client.

Reply:
> EDIT: I tried to fetch code from `header.parent_hash()` but it doesn't seem to work either

This should work.

Reply:
indeed I mixed up, I triggered a successful runtime upgrade.
I fail to trigger a second one, but I think this is because of some other constraint I'm not aware of.